drivers/edac/edac_ibecc.c

```diff
@@ -0,0 +1,457 @@
+
+static uint64_t inject_addr_mask_get(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	return sys_read64((mem_addr_t)((uint8_t *)data->mchbar +
+				       IBECC_INJ_ADDR_MASK));
+}
+
+static int inject_ctrl_set(const struct device *dev, uint32_t ctrl)
+{
+	struct ibecc_data *data = dev->data;
+
+	sys_write32(ctrl, (mem_addr_t)((uint8_t *)data->mchbar +
+				       IBECC_INJ_ADDR_CTRL));
+
+	return 0;
+}
+#endif /* CONFIG_EDAC_ERROR_INJECT */
+
+static uint64_t ecc_error_log_get(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	LOG_DBG("Get ECC Error Log");
+
+	return sys_read64((mem_addr_t)((uint8_t *)data->mchbar +
+				       IBECC_ECC_ERROR_LOG));
+}
+
+static void ecc_error_log_clear(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	LOG_DBG("Clearing ECC Error Log");
+
+	/* Clear all error bits */
+	sys_write64(ECC_ERROR_MERRSTS | ECC_ERROR_CERRSTS,
+		    (mm_reg_t)((uint8_t *)data->mchbar + IBECC_ECC_ERROR_LOG));
+}
+
+static uint64_t parity_error_log_get(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	LOG_DBG("Get Parity Error Log");
+
+	return sys_read64((mem_addr_t)((uint8_t *)data->mchbar +
+				       IBECC_PARITY_ERROR_LOG));
+}
+
+static void parity_error_log_clear(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	LOG_DBG("Clearing Parity Error Log");
+
+	sys_write64(PARITY_ERROR_ERRSTS,
+		    (mm_reg_t)((uint8_t *)data->mchbar +
+			       IBECC_PARITY_ERROR_LOG));
+}
+
+static unsigned int errors_cor_get(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	return data->errors_cor;
+}
+
+static unsigned int errors_uc_get(const struct device *dev)
+{
+	struct ibecc_data *data = dev->data;
+
+	return data->errors_uc;
+}
+
+static int notify_callback_set(const struct device *dev,
+			       edac_notify_callback_f cb)
+{
+	struct ibecc_data *data = dev->data;
+	int key = irq_lock();
+
+	data->cb = cb;
+	irq_unlock(key);
+
+	return 0;
+}
+
+static const struct edac_driver_api api = {
+#if defined(CONFIG_EDAC_ERROR_INJECT)
+	/* Error Injection functions */
+	.inject_addr_set = inject_addr_set,
+	.inject_addr_get = inject_addr_get,
+	.inject_addr_mask_set = inject_addr_mask_set,
+	.inject_addr_mask_get = inject_addr_mask_get,
+	.inject_ctrl_set = inject_ctrl_set,
+#endif /* CONFIG_EDAC_ERROR_INJECT */
+
+	/* Error reporting & clearing functions */
+	.ecc_error_log_get = ecc_error_log_get,
+	.ecc_error_log_clear = ecc_error_log_clear,
+	.parity_error_log_get = parity_error_log_get,
+	.parity_error_log_clear = parity_error_log_clear,
+
+	/* Get error stats */
+	.errors_cor_get = errors_cor_get,
+	.errors_uc_get = errors_uc_get,
+
+	/* Notification callback set */
+	.notify_cb_set = notify_callback_set,
+};
+
+int edac_ibecc_init(const struct device *dev)
+{
+	const pcie_bdf_t bdf = PCIE_BDF(0, 0, 0);
+	struct ibecc_data *data = dev->data;
+	uint32_t tolud;
+	uint64_t touud, tom, mchbar;
+
+	ARG_UNUSED(dev);
```

Comment:
useless, dev is beitg used above to retrieve the private data.

Reply:
I will remove this.